demo/future/spec.md

```diff
@@ -0,0 +1,635 @@
+# Formatic 3000
```

Comment:
@jdeal I took a very quick glance at this spec, but not the code. None of this seems problematic with how the design system is building components since largely this new version of formatic allows customizing how any form component renders 🎉

Overall this spec looks great to me. Here are some questions / feedback that I have:

1. There are a lot of examples and the API surface area still seems very large and somewhat ambiguous to me. Details below on individual points.
1. Is there a benefit of providing `FieldContainer` _and_ hooks, as in does each solve unique problems that the other cannot? If they provide equivalent functionality then IMO there should only be one. If there's only one option then there's no choice to make, which I think is a good thing.
1. I don't follow the difference between `renderTag` and `renderComponent`. Does there need to be an actual difference between the two or could there only be one? I would expect only `renderComponent` to be necessary because I would think the rendered components would control their own DOM, so I don't follow where `renderTag` fits in and when it would be called.
1. I don't understand `renderKey` and am curious if it's necessary. I would assume that formatic would choose which component to render for me given the type of the field. How it compares to `field.type` is also confusing to me. If it _is_ necessary, are there a finite number of values that `renderKey` could be? IMO it would make sense to provide those as constants rather than strings. Perhaps `Symbol`s so you _must_ import them and can't shoot yourself in the foot by having a lot of strings throughout the codebase.
1. Do nested values require nesting components or does/could `fieldKey` support something like `my.nested.value` or even `['my', 'nested', 'value']`?
1. The `AutoFields` examples are a little hard for me to understand at a glance but seem to provide the functionality I would expect based on my above points. For instance mapping `String: MyStringField` and `Integer: MyIntegerField` makes a lot of sense to me. However I don't understand how `AutoFields` either complements or competes with `renderTag` and `renderComponent` or `AutoFieldsContainer`.

Overall it seems like there's a lot of API flexibility that is provided, but I think it makes it ambiguous which API to choose and how they compete with each other. If it comes down to personal preference rather than actual tradeoffs of one API vs another, I think formatic should choose one to minimize its API. I definitely think less is more here.

Reply:
@stevenhauser thanks for taking a look!

1. Hmm, I like to think the API surface is a lot simpler than the current version of Formatic, where the plugins have a lot of hidden surface area. But I'll answer specific points...
2. I see what you're saying, but I think that's a userland decision. Sometimes, it might make sense to inline a render callback because you have a bunch of variables in scope and you want to dynamically render a field based on those. Being forced to use `useField` would mean you have to pass all those variables as props in a case where it's not very ergonomic. Also, this makes it easy to make dependent fields where a field only shows up based on the value of another field. I'll likely create something higher level for that, but right now, I'm focusing on the primitives that you can stitch together. I appreciate wanting to keep the API small, and I know there's some overlap, but I think it's worth keeping the render callback for now. After I build more examples, I can revisit though.
3. I originally tried to do a single callback, but it gets pretty weird in practice. `renderTag` is specifically meant for rendering DOM nodes. Although it will probably be recommended that userland do all the DOM rendering, I think it's nice to have reference components to quickly build a form and see if Formatic works vs having to build custom components right away. `renderTag` exists mainly as a way to style the DOM elements of those reference components. `renderComponent` is used to completely take over the rendering of a component. If you take over all component rendering, you don't need to worry about `renderTag` at all.
4. I'm building the automatic form generation from other primitives. So you can have component/tag overriding independent of automatic form generation. I started out with symbols actually, but it felt kind of overkill. I'd be fine putting it back in though.
5. I imagine I'll add some sugar, but I wanted to focus on primitives first. I'd definitely want to avoid having to always parse the keys, so I'd want to make that a different prop like `keyPath` or something.
6. `AutoFields` mainly exists for you to map automatic form generation to _your_ components. If you use built-in Formatic components, you'll be able to use `renderComponent` and `renderTag` on those, but if you use your own components, you would not (since you wouldn't have built your components to be overridable unless Formatic exports a utility for that).

Hopefully that helps explain some things @stevenhauser ! I kind of got lazy with the spec at the end, so as I flesh out more examples, it may make more sense. Or, maybe it will become more obvious what can be cut. Let me know if something I've said here still doesn't make sense though! (Happy to chat sync too.)

Reply:
Here's some more nuance worth knowing about `renderComponent`, `renderTag`, `AutoFields`, and `AutoFieldsContainer`:

`renderComponent` is meant for _intercepting_ or _overriding_ the rendering of components that would be rendered anyway if you did _nothing_. This means that you've already absorbed the cost of those components in the build, so you should either be intercepting to modify/add props or lightly overriding small components. You would not want to use `renderComponent` to _completely replace_ a bunch of components, since you'd be including all those components in the build and then just replacing them anyway. This is a problem today with Formatic. All components are part of the default config, so even if you override the rendering of every single component, you're still paying the tax for all those default components.

That's one reason it's convenient to separate `renderTag`: with `renderTag`, you're rendering DOM nodes, so you want to render _every single bit_ of the component. You just want to modify/add DOM-related props. Without that separation, you'd have to have a conditional inside your `renderComponent` function to determine what you're rendering. A `data-*` attribute makes sense in the case of a DOM-specific component, but not a DOM-independent component, for example.

`AutoFields` is a kitchen-sink component. It likely will exist simply to take Formatic for a test-drive or for consumers who don't care too much about presentation (like for admin pages) and just want to quickly dump a form to the page. It's going to absorb all of Formatic's built-in components into your build (like with today's Formatic). You don't want to use `AutoFields` and then override everything with `renderComponent` (although you totally could do that).

`AutoFieldsContainer` on the other hand takes a `components` prop so you can pass in your own components. Or you could pick and choose to import only those components you want from Formatic's built-in components vs importing the kitchen-sink with `AutoFields`. We can't have a single `AutoFields` component that does both these things, because once we've imported everything, we can't unimport everything. Tree-shaking has no idea which types you're going to pick at runtime.

Now, it does _feel_ like the `components` prop of `AutoFieldsContainer` and `renderComponent` are doing similar jobs. But look at the signature of `renderComponent`:

```js
renderComponent(renderKey, Component, props)
```

Right now, it's meant for _overriding_ `Component`. But in the case of `AutoFieldsContainer`, we don't actually have a component to override. That's the point: we want to decide what to render in userland. If we did that, we'd be calling `renderComponent` with something like this:

```js
renderComponent('Text', undefined, {...});
```

Maybe that's fine? It's a little weird though because it means that function is doing double-duty and like with the problem of combining `renderTag`, you'd have to have a conditional in there in some cases. And the documentation would have to explain that subtle/strange thing.

We could also have a third function like `renderAutoFieldComponent` or whatever, which would just be called like:

```js
renderAutoFieldComponent('Text', {...});
```

That's clean and is a little more flexible than the `components` property since you can do some dynamic wrapping if you need to. However, that case is atypical and didn't seem to warrant the extra complexity vs just having an object pointing to components. If you really need to inject something dynamic, you can pretty easily do that with context and a hook. The reverse is true as well. You could use an object mapping instead of `renderComponent`, but there, we're talking about potentially many more layers of wrapping if you needed to do something dynamic, which might run into perf problems. (And there's already some wrapping cost to make the components overridable. I tried to avoid that, but that's another story of many traps.) So I tried to choose pragmatically for each use case.

Sorry about the wall of text! This is as much me thinking out loud as anything. So feel free to ignore. But happy to hear if you have any alternate design ideas.